### ISSUE TITLE
Scope Manager fails to pick up references to implicit globals

### ISSUE THREAD

**Issue by lilnasy** [typescript-eslint/typescript-eslint#11714]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug) and found none that matched my issue.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Relevant Package

scope-manager

### Repro Code

```TypeScript
undefined; globalThis; NaN; Object; Boolean; String; Math; Date; Array; Map; Set; var foo; foo;
```

### ESLint Config

```javascript
Not relevant
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "lib": ["esnext"]
  }
}
```

### Expected Result

All identifiers but `foo` are analyzed to be determined references to implicit global variables.

### Actual Result

Only ~~`String`~~ `Boolean` and `Object` are determined to be references to implicit global variables.

### Additional Info

- I encountered this while working on https://github.com/oxc-project/oxc/pull/14890.
### What's going wrong?
- The generated code for ts libs may include multiple mentions of the same variable, in some cases indicating the variable is a [`TYPE`](https://github.com/typescript-eslint/typescript-eslint/blob/55ca033ee88cc95cfac4ad6dea2257fbeb1d4657/packages/scope-manager/src/lib/base-config.ts#L16), and `TYPE_VALUE` in others.
- When `Referencer.populateGlobalsFromLib` calls`GlobalScope.defineImplicitVariable` which calls `ScopeBase.defineVariable` and the variable is already defined, it doesn't entertain the possibility that the variable may have different type/value attributes and simply moves on.
- The result: `@typescript-eslint/scope-manager` incorrectly indicates to the user project that a variable is a type, and the scope analysis does not pick up value identifiers as references to the variable.

### Versions

| package                                | version |
| -------------------------------------- | ------- |
| `@typescript-eslint/scope-manager`     | `8.46.1'` |
| `TypeScript`                           | `5.9.3` |
| `node`                                 | `24.10.0` |

**Comment by JoshuaKGoldberg** [comment]
Hey thanks for using the scope-manager package & reporting this @lilnasy! To be honest, we don't touch the package often, so we'll need to be careful in changing nuances such as the implicit global references.

Just to confirm we're on the same page, I think what you're saying is that most of the names in the list (`Array`, `Date`, `Map`, etc.) aren't set as `isValueVariable`, though they should? That only `Object` and `String` are?

I pushed one up here of what I think is going on: https://github.com/JoshuaKGoldberg/repros/tree/tseslint-scope-manager-implicit-globals. In it, it's `Object` and _`Boolean`_, not `String`, that have `isValueVariable` unlike the others. Are we looking at different things?

**Comment by lilnasy** [comment]
The difference may be due to the configured lib. I had it configured to esnext.

Edit: just remembered esnext is the default, so that's not it. I may be misremembering. Regardless of the exact globals, the issue is the same.

**Comment by bradzacher** [comment]
If I am understanding correctly -- I think that you might be getting bitten by some intentional behaviour which we had to implement to align with ESLint core.
ESLint itself expects that it will need to resolve the unresolved global references at the start of the lint run:
https://github.com/eslint/eslint/blob/fef04b5c7fea99362d67b31b8e98cd4914020ed3/lib/linter/linter.js#L232-L253

This will likely change soon -- see https://github.com/eslint/js/issues/665

If you have a look [at the `Scope` tab in our playground](https://typescript-eslint.io/play/#ts=5.9.2&showAST=scope&fileType=.tsx&code=K4OwJgpgZgliFgNwAIDmAbA9gIwIboBUALGAZxQDlcKUB5bAKwgGMAXFAIU03QlxBQBlVgCc4qFAFlcrIigAiMiCgCCIkbgCeU3AAchEdsgBuuEcijcUlzIgBQQA&eslintrc=N4KABGBEBOCuA2BTAzpAXGYBfEWg&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false) you'll see that the references are all resolved there (because the info there is the scope info after eslint's behaviour has been applied).

<img width="533" height="995" alt="screenshot of playground showing scope info" src="https://github.com/user-attachments/assets/47f28f34-86a6-4c32-bae2-e3cc11f7b3c1" />

Please correct me if I've misunderstood.

**Comment by lilnasy** [comment]
@bradzacher Thanks for looking into it!

That does indeed seems intentional in `escope` according to this comment from 2015: https://github.com/eslint/eslint/issues/4615#issuecomment-162276435. However, I am not sure how well that translates to `@typescript-eslint/scope-manager` since statically available globals are known to a good extent unlike with `escope`.

At the very least, I would find it surprising if `isValueVariable` being `false` is considered intentional for most global value variables, which I can see in the playground you linked as well.

**Comment by overlookmotel** [comment]
Looks like making this change is proposed due to upcoming changes in ESLint v10: #11762

**Comment by JoshuaKGoldberg** [comment]
Blagh, I think this was meant to be put as `status: accepting prs` a while ago. Sorry about that.

Note @bradzacher's https://github.com/typescript-eslint/typescript-eslint/pull/11715#pullrequestreview-3413879512
